### PR TITLE
Added possibility to reset the cameras at startup.

### PR DESCRIPTION
**Pull request by S-Dafarra**
With this option, the module tries to reset the cameras at startup. It can be disabled from configuration files.

### CODE REVIEW

File: modules/Oculus_module/src/OculusModule.cpp [hunk 1]
```diff
@@ -416,6 +418,73 @@ bool OculusModule::configure(yarp::os::ResourceFinder& rf)
...
+
+        if (!YarpHelper::getStringFromSearchable(generalOptions, "rightCameraPort", rightCamera))
+        {
+            yError() << "[OculusModule::configure] resetCameras is non-zero, but no rightCameraPort is provided.";
+            return false;
+        }
+
+        if (!cameraReset(leftCamera,"/walking-teleoperation/camera-reset/left"))
+        {
+            yError() << "Failed to reset left camera.";
```
Comment (GiulioRomualdi) [inline modules/Oculus_module/src/OculusModule.cpp:476]:
```suggestion
            yError() << "[OculusModule::configure] Failed to reset left camera.";
```

Comment (GiulioRomualdi) [reply modules/Oculus_module/src/OculusModule.cpp:476]: also for the other prints

Comment (S-Dafarra) [reply modules/Oculus_module/src/OculusModule.cpp]: Good point. I added it in https://github.com/robotology/walking-teleoperation/pull/45/commits/4f76580403f63782e5611cdc0326a2f2649b33c5

**Review by GiulioRomualdi — APPROVED** [review]
Thank you @S-Dafarra for working on this. It's really appreciated 

Please @kouroshD merge the PR if you are fine with the modifications

File: modules/Oculus_module/src/OculusModule.cpp [hunk 2]
```diff
@@ -416,6 +418,73 @@ bool OculusModule::configure(yarp::os::ResourceFinder& rf)
     }
     m_oculusHeadsetPoseInertial.resize(6, 0.0);
 
+    //Reset the cameras if necessary
+    bool resetCameras = generalOptions.check("resetCameras", yarp::os::Value(false)).asBool();
+    yInfo() << "[OculusModule::configure] Reset camera: " << resetCameras;
+    if (resetCameras)
+    {
+        auto cameraReset = [](const std::string& cameraPort, const std::string& localPort) -> bool
```
Comment (kouroshD) [inline modules/Oculus_module/src/OculusModule.cpp:426]: Is there any reason you decided to use lambda function instead of a private member function of the class? I a saying that, since the code of this function is inside `configure` member function, and the code here is long in this function and adding this lambda function make it longer even. Also, it reduces the readability.
In any case, in line with other private functions we have such as `configureJoypad` and `configureOculus` please use `configureCamera` private member function.

**Review by kouroshD — CHANGES_REQUESTED** [review]

**Comment by kouroshD** [comment]
Thank you @S-Dafarra  for the PR. I had some comments and requests which I have mentioned them with the review.

File: app/robots/iCubGenova04/oculusConfig.ini [hunk 3]
```diff
@@ -22,6 +22,12 @@ playerOrientationThreshold    0.2
 humanHeight                   1.76
 # The robot arm span is the distance between the left and right index fingertips when the arms are aligned, i.e. T pose.
 robotArmSpan                  1.10
+# If resetCameras !=0, the module will attempt to reset the cameras during the initialization phase.
```
[the remark this answers is not in the document]
Comment (S-Dafarra) [reply app/robots/iCubGenova04/oculusConfig.ini:25]: Well, in principle that's not true. As a matter of fact, the value is a char, since it is in a text file. It is then interpreted as a boolean. This means that every value different from zero will be considered ``true``. So, 
> ``If resetCameras =1, the module will attempt to reset the cameras during the initialization phase.``

would not be precise, as also ``resetCameras =7`` would have the same effect.

File: app/robots/iCubGenova04/oculusConfig.ini [hunk 4]
```diff
@@ -22,6 +22,12 @@ playerOrientationThreshold    0.2
...
+resetCameras            1
+# The leftCameraPort and rightCameraPort are needed only if resetCameras is 1
+leftCameraPort          /icub/cam/left
+rightCameraPort         /icub/cam/right
+
```
[the remark this answers is not in the document]
Comment (S-Dafarra) [reply app/robots/iCubGenova04/oculusConfig.ini:30]: Nope, we have to connect to the cameras directly. Indeed, I need to open the camera driver, and to do so I need to know the camera remote.

File: app/scripts/dcmWalkingRetargeting.xml [hunk 5]
```diff
@@ -58,6 +58,8 @@
     <dependencies>
       <port timeout="5.0">/transformServer/transforms:o</port>
       <port timeout="5.0">/joypadDevice/Oculus/rpc:i</port>
+      <port timeout="5.0">/icub/cam/left</port>
```
[the remark this answers is not in the document]
Comment (S-Dafarra) [reply app/scripts/dcmWalkingRetargeting.xml:61]: I added these lines since the resetting of the cameras occurs as soon as we start the module. If ``resetCameras`` is enabled, it needs to connect to these ports immediately.

File: modules/Oculus_module/src/OculusModule.cpp [hunk 6]
```diff
@@ -12,13 +12,17 @@
 #include <yarp/os/LogStream.h>
 #include <yarp/os/Property.h>
 #include <yarp/os/Stamp.h>
+#include <yarp/dev/FrameGrabberInterfaces.h>
```
[the remark this answers is not in the document]
Comment (S-Dafarra) [reply modules/Oculus_module/src/OculusModule.cpp:15]: Nope. I am just adding the interface.  That file should be always installed by ``yarp``. The respective device instead is the ``dragonfly2``, that should be enabled in order to have the cameras working. 
In our specific case, it is enabled with https://github.com/robotology/robotology-superbuild/blob/master/cmake/BuildICUB.cmake#L53
and then it implements the ``FrameGrabberInterface`` that I need  https://github.com/robotology/icub-main/blob/5d491cf71ac08dc0e7774c81a5dd0f12035af7c6/src/libraries/icubmod/dragonfly2/common/DragonflyDeviceDriver2.h#L291

So, bottom line is that as long as the cameras on the robot are working (and are dragonfly), that should work.

File: modules/Oculus_module/src/OculusModule.cpp [hunk 7]
```diff
@@ -215,8 +219,6 @@ bool OculusModule::configure(yarp::os::ResourceFinder& rf)
     yInfo() << "[OculusModule::configure] matlogger2 is disabled!";
 #endif
 
-    yarp::os::Value* value;
```
[the remark this answers is not in the document]
Comment (S-Dafarra) [reply modules/Oculus_module/src/OculusModule.cpp:218]: It seems it was not used at all.

File: modules/Oculus_module/src/OculusModule.cpp [hunk 8]
```diff
@@ -416,6 +418,73 @@ bool OculusModule::configure(yarp::os::ResourceFinder& rf)
...
+            {
+                yError() << "[OculusModule::configure] Cannot open remote_grabber device on port " << cameraPort <<".";
+                return false;
+            }
+
+            yarp::dev::IFrameGrabberControlsDC1394 *grabberInterface;
+
+            if(!grabberDriver.view(grabberInterface) || !grabberInterface)
+            {
+                yWarning() << "[OculusModule::configure] RemoteGrabber does not have IFrameGrabberControlDC1394 interface, please update yarp.";
```
[the remark this answers is not in the document]
Comment (S-Dafarra) [reply modules/Oculus_module/src/OculusModule.cpp]: Done in https://github.com/robotology/walking-teleoperation/pull/45/commits/16413ff242361bb6c487e4d941bcee5a7bfcbb2e

File: modules/Oculus_module/src/OculusModule.cpp [hunk 9]
```diff
@@ -416,6 +418,73 @@ bool OculusModule::configure(yarp::os::ResourceFinder& rf)
...
+                return false;
+            }
+
+            if(!grabberInterface->setResetDC1394())
+            {
+                yWarning() << "[OculusModule::configure] Failed to reset the camera on port " << cameraPort << ".";
```
[the remark this answers is not in the document]
Comment (S-Dafarra) [reply modules/Oculus_module/src/OculusModule.cpp]: Done in https://github.com/robotology/walking-teleoperation/pull/45/commits/16413ff242361bb6c487e4d941bcee5a7bfcbb2e

File: modules/Oculus_module/src/OculusModule.cpp [hunk 10]
```diff
@@ -416,6 +418,73 @@ bool OculusModule::configure(yarp::os::ResourceFinder& rf)
...
+
+            grabberDriver.close();
+
+            return true;
+        };
+
+        std::string leftCamera, rightCamera;
+        if (!YarpHelper::getStringFromSearchable(generalOptions, "leftCameraPort", leftCamera))
+        {
+            yError() << "[OculusModule::configure] resetCameras is non-zero, but no leftCameraPort is provided.";
```
[the remark this answers is not in the document]
Comment (S-Dafarra) [reply modules/Oculus_module/src/OculusModule.cpp]: Done in https://github.com/robotology/walking-teleoperation/pull/45/commits/16413ff242361bb6c487e4d941bcee5a7bfcbb2e

File: modules/Oculus_module/src/OculusModule.cpp [hunk 11]
```diff
@@ -416,6 +418,73 @@ bool OculusModule::configure(yarp::os::ResourceFinder& rf)
...
+            return false;
+        }
+
+        if (!YarpHelper::getStringFromSearchable(generalOptions, "rightCameraPort", rightCamera))
+        {
+            yError() << "[OculusModule::configure] resetCameras is non-zero, but no rightCameraPort is provided.";
```
[the remark this answers is not in the document]
Comment (S-Dafarra) [reply modules/Oculus_module/src/OculusModule.cpp]: Done in https://github.com/robotology/walking-teleoperation/pull/45/commits/16413ff242361bb6c487e4d941bcee5a7bfcbb2e

File: modules/Oculus_module/src/OculusModule.cpp [hunk 12]
```diff
@@ -416,6 +418,73 @@ bool OculusModule::configure(yarp::os::ResourceFinder& rf)
...
+                yWarning() << "[OculusModule::configure] Failed to reset the camera on port " << cameraPort << ".";
+                return false;
+            }
+
+            grabberDriver.close();
+
+            return true;
+        };
+
+        std::string leftCamera, rightCamera;
```
[the remark this answers is not in the document]
Comment (S-Dafarra) [reply modules/Oculus_module/src/OculusModule.cpp]: Done in https://github.com/robotology/walking-teleoperation/pull/45/commits/16413ff242361bb6c487e4d941bcee5a7bfcbb2e

File: modules/Oculus_module/src/OculusModule.cpp (hunk 2 above)
Comment (S-Dafarra) [reply modules/Oculus_module/src/OculusModule.cpp]: Moved to a method ``resetCamera`` (not ``configure`` since I am not really configuring anything) in https://github.com/robotology/walking-teleoperation/pull/45/commits/16413ff242361bb6c487e4d941bcee5a7bfcbb2e

File: app/robots/iCubGenova04/oculusConfig.ini (hunk 3 above)
[the remark this answers is not in the document]
Comment (kouroshD) [reply app/robots/iCubGenova04/oculusConfig.ini:25]: OK, make sense. So, in `    bool resetCameras = generalOptions.check("resetCameras", yarp::os::Value(false)).asBool();` if the char value is not `0`, the behavior of the function instantiate `resetCameras` with true value, right? @S-Dafarra

File: modules/Oculus_module/src/OculusModule.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (kouroshD) [reply modules/Oculus_module/src/OculusModule.cpp:15]: OK, I see, thanks for explanation.
So, just to make the path correctly, we have an option `ROBOTOLOGY_ENABLE_ICUB_HEAD ` in superbuild. If this is `on`, then the icub module also installs the device as you pointed out in https://github.com/robotology/robotology-superbuild/blob/master/cmake/BuildICUB.cmake#L53 and we have this option only on ubuntu machine of the icub. Then my question is, what if we are not running the oculus module code without iCub ? What will be the output of `resetCamera` function, considering opening the device and the result of `grabberInterface->setResetDC1394()`?

File: app/robots/iCubGenova04/oculusConfig.ini (hunk 3 above)
[the remark this answers is not in the document]
Comment (S-Dafarra) [reply app/robots/iCubGenova04/oculusConfig.ini:25]: Yes, as per the [documentation](http://www.yarp.it/git-master/classyarp_1_1os_1_1Value.html#a3f49bd86a14537ca63f59297513e80eb)
> boolean value if value is indeed a boolean. If it is another type, the appropriate cast value is returned. Otherwise returns false.

and basically, everything that is different from 0 or ``false`` is casted to ``true``.

File: modules/Oculus_module/src/OculusModule.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (S-Dafarra) [reply modules/Oculus_module/src/OculusModule.cpp:15]: Not sure I understood your concern
In any case:
- the module compiles even if the option ``ROBOTOLOGY_ENABLE_ICUB_HEAD`` is set to ``OFF``. In fact, I tested it also by running the ``OculusRetargetingModule`` on my PC.
- If you are not connected to any robot, the configuration will fail before because of the other retargeting modules.
- If you are connected to a robot, and for some reason you don't want to use the eyes, you can simply set the option ``resetCameras`` to 0. Indeed, by default ``resetCameras`` is false, and I added this parameter only to the ``iCubGenova04`` configuration files. 
- Any problems related to the opening of the camera driver would occur only if  ``resetCameras`` is non-zero.

[the remark this answers is not in the document]
Comment (kouroshD) [reply modules/Oculus_module/src/OculusModule.cpp:15]: OK, that sounds good. So we cover different situations without causing any problem. Thanks.

**Review by kouroshD — APPROVED** [review]

**Comment by kouroshD** [comment]
Thank you @S-Dafarra  for this PR, confirmed and merged.
